### DISCUSSION TITLE
Using the status page with a private repo

### DISCUSSION THREAD

**Discussion by akoenig** [upptime/upptime#54] [Ideas] [answered]
Really awesome idea! Thanks for building it 🙂 🚀 

I have one question though. The frontend (status page) is requesting data fron the GitHub API – in order to fetch `history/summary.json` and the incident comments, etc. This results in a case where `upptime` can't be used within a private repository setting where the status page gets still deployed to GitHub Pages. Do you think that this might be a use-case which `upptime` should / will cover in the future or is it already possible and I'm missing something here? 🙂

**Comment by AnandChowdhary** [comment]
No, it's not possible (at this time) to have a status website with data from a private repository. There is probably a way that we can call the GitHub API on built time and regenerate a static website every time there is a new incident, but it'll be hard to keep track of every new comment on an issue. IMHO, it's just better to call the API since the data will always be real-time, so this is not a planned feature.

tl;dr: Not possible for now, no concrete plans to add this functionality.

> If you don't want to have a publicly-available status website, you don't have to enable publishing. If that is the case, you can also choose to keep your repository private. https://upptime.js.org/docs/get-started

**Comment by AnandChowdhary** [comment]
If someone wants to send a PR for this feature, I'll reopen this issue.

**Comment by akoenig** [comment]
Ah, I see. No worries and thanks for the reply. One option which I have in mind is to have a GitHub-compatible API which proxies the needed requests accordingly.

Just saw that the status-page is using octokit. Not sure if it allows to change the API host though.

**Comment by AnandChowdhary** [comment]
Yes, you can specify the `baseUrl` like so: https://octokit.github.io/rest.js/v18#usage

Maybe I can add a configuration option for the base URL, which you can then use to proxy API requests with a personal access token with readonly access to the repository. What do you think?

**Comment by akoenig** [comment]
That would be superb @AnandChowdhary!!

Just figured that we have to find a solution for the Graphs as well, right?

**Comment by akoenig** [comment, marked as answer]
@AnandChowdhary I opened a PR which introduces the configuration 🙂 Due to the fact that the graphs are fetched via raw.githubusercontent.com, I had to make sure that when a different API Base URL has been configured, the respective proxy needs to tunnel those requests as well. Luckily, the same authorization mechanism can be used by passing the PAT as an `Authorization` header.

**Reply by boredland** [reply]
Wouldn't it be nicer if we added all the files the page needs to the gh-pages branch and not need all of this?

**Comment by AnandChowdhary** [comment]
This is great, thanks!

**Comment by akoenig** [comment]
@AnandChowdhary You're welcome 🙂

Anything I should have in mind when I want to use it? Just created a fresh `upptime` repository from the template and configured the `apiBaseUrl`. At the moment, it still communicates with `api.github.com`.

**Edit:** Seems like the `status-page` package needs a new release and being updated in `uptime-monitor` or is this the wrong path? 🙂

**Comment by AnandChowdhary** [comment]
Yes, you're right, since you didn't use gitmoji or conventional commits, the automated semantic release didn't trigger. No worries, I've gone ahead and manually added a new release for you to use. 🙂

New release: https://github.com/upptime/status-page/releases/tag/v1.4.0

**Comment by akoenig** [comment]
Awesome, thanks @AnandChowdhary. Ah okay, sorry. I thought I followed your convention and was wondering why there is no conventional commit message in place, but it pointed out that I just overlooked it obviously 😅

**Comment by AnandChowdhary** [comment]
No worries, thanks again for your contribution! 😄

**Comment by AnandChowdhary** [comment]
@akoenig, there's a breaking change you should be aware of: In https://github.com/upptime/upptime/issues/107, we decided that we shouldn't publish status websites for private repos by default. I think this is a better default because if you have all private data, exposing it by default isn't the best idea.

For people who have a private repo and still want to publish their status website, they now have to set this:

```yaml
status-website:
  publish: true
```

**Reply by pdellarciprete** [reply]
It is not documented :(

**Reply by AnandChowdhary** [reply]
Sorry about that! Please feel free to create a pull request!

**Comment by akoenig** [comment]
@AnandChowdhary Ah great, thanks for the info. 😊

Introducing this attribute makes sense 💪

**Comment by Tragio** [comment]
Hello community! 🎄 

I was trying Upptime with a private repo and a custom domain, however not sure what I'm doing wrong. When I access the website I see the dashboard for a second and then it goes to an error page. I'm using the following:

```yaml
status-website:
  publish: true
  apiBaseUrl: ${{ secrets.GH_PAT }}
  cname: subdomain.domain.com
```

I'm using the same token as the GH_PAT since it also has the repo permissions. Not sure if I should or what I'm doing wrong. I also tried using the plain token on the configuration file but the same problem.

![Upptime](https://user-images.githubusercontent.com/5109593/103087145-71d87700-45de-11eb-8683-ff4b843d693e.jpg)

Really thank you for this amazing work! 😃

**Reply by AnandChowdhary** [reply]
@Tragio:

> I was trying Upptime with a private repo and a custom domain, however not sure what I'm doing wrong. When I access the website I see the dashboard for a second and then it goes to an error page. I'm using the following:
> 
> ```yaml
> status-website:
>   publish: true
>   apiBaseUrl: ${{ secrets.GH_PAT }}
>   cname: subdomain.domain.com
> ```

You're setting the PAT as your API base URL, when you should set up a proxy server and add that as the URL. Maybe @akoenig can elaborate if he has a quick tool to add the authentication header and forward requests to the GitHub API.

**Reply by Krakaw** [reply]
I used an Nginx reverse proxy to proxy the requests, and put a basic auth wall in front of that using https://github.com/Krakaw/rauthy

```nginx
server {
    server_name my-gh-proxy.example.com;
    listen 80;
    location = /rauthy {
        internal;
        proxy_pass                          http://127.0.0.1:3031$request_uri;
        proxy_pass_request_body             off;
        proxy_set_header                    Content-Length "";
        proxy_set_header X-Real-IP          $remote_addr;
        proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
        proxy_pass_request_headers          on;
    }
    location / {
        auth_request /rauthy;
        proxy_set_header Authorization     "token $YOUR_GH_PAT";
        proxy_pass https://api.github.com;
    }
}
```

**Reply by akoenig** [reply]
Sorry for the late response. Yap, the solution @Krakaw suggested is exactly the approach we have established as well. One addition though: I created the PAT (with repo read scope) for a user which is exclusively added to the private Upptime repo.

**Reply by mig5** [reply]
Can you share more about this solution? I tried rauthy but I can't seem to get it to work, I just keep getting a 401 auth popup... I've set the GH_PAT in the Authorization header..

Is there meant to be more proxy_pass stuff to pass to the backend Github Pages URL ?

**Reply by flauc** [reply]
This doesn't seem to be working. For starters you would need to proxy to both https://api.github.com and https://raw.githubusercontent.com, because the site makes calls to both. Additionally it can't be just forwarding the requests since depending on the route you need to send it to one of the two places. I was able to get it working with the below server: 

```
const express = require('express');
const { createProxyMiddleware } = require('http-proxy-middleware');

const app = express();

app.use('/status-board/[my-organization-name]', createProxyMiddleware({
  target: `https://raw.githubusercontent.com`,
  changeOrigin: true,
  pathRewrite: {[`^/status-board/`]: '/'},
  onProxyReq(proxyReq, req, res) {
    proxyReq.setHeader('Authorization', 'Bearer [my-token]');
  }
}));

app.use('/status-board/repos', createProxyMiddleware({
  target: `https://api.github.com`,
  changeOrigin: true,
  pathRewrite: {[`^/status-board/`]: '/'},
  onProxyReq(proxyReq, req, res) {
    proxyReq.setHeader('Authorization', 'Bearer [my-token]');
  }
}));

app.listen(3000, () => {
  console.log('Server listening on port ', 3000);
});
```

This is the full source code, I just replaced the token and organization. This is what I added as the `apiBaseUrl: https://my-server.com/status-board`

**Reply by mig5** [reply]
@flauc . Thanks. I tried your nodeJS verbatim (adjusting just the token), along with updating the apiBaseUrl, and I just get `Cannot GET /status-board`

This is starting to seem surprisingly hard to set up...

**Reply by flauc** [reply]
@mig5 this does work on my end. What's the full link to your server? Did you adjust this part `app.use('/status-board/[my-organization-name]', createProxyMiddleware({`

**Reply by mig5** [reply]
@flauc thanks for helping me. I did indeed adjust that part.

My site is at https://uptime.mixcic.net (you can't access it, it's behind a VPN on an internal network)

I've made a Gist [here](https://gist.github.com/mig5/84020f337a43fc8f3104e654e6104b6d) that shows:

a) The upptime.js file I made (copying yours). I am running this with `node upptime.js` (node v10.23.2)
b) The nginx vhost that sits in front, for HTTPS
c) The snippet from my .upptimerc.yml that shows the apiBaseUrl

I am confident that it is not the Nginx vhost, because even hitting the Node service at port 3000 throws the same `Cannot GET /status-board` message. I wonder if there are some debug statements I can add? I'm a NodeJS novice, unfortunately. 'Cannot GET' is a rather vague term, hard to tell if it's some network issue, issue with the token, issue on the Github side with my upptime repo, or what..

**Reply by flauc** [reply]
Hey @mig5, I'm sad to hear you're still struggling with it, however, this seems to be a network issue. You can try this locallly when running with `node upptime.js`. Here is an example request that uptime uses, it should either return results or a 404 json response (so not the html 404 that is a default express response when a route doesn't match) : 

http://localhost:3000/status-board/Mydex/system-monitor/master/history/summary.json.

If that works, then it's definitely something downstream that's not working properly, and if you get a json 404, then it's probably a bad token (you should create a dummy user with read access only to the upptime repo and create a token through him).

**Reply by mig5** [reply]
@flauc thanks - I was able to see https://uptime.mixcic.net/status-board/Mydex/upptime/master/history/summary.json  . I think part of the confusion was my repo is called 'upptime' where as yours has been named 'system-monitor' above.

But the other issue was that I ended up needing to make sure that I had the `/upptime` as my `baseUrl` (I thought I needed the CNAME still) and therefore to still access the site as https://mydex.github.io/upptime/ . There are certain other quirks such as it doesn't work if I omit the trailing slash at the end of the URL, and the graphs don't appear if I add /index.html . Rather finicky! But it works!

In short - mostly just confusion on my part.

Now it all works! Thanks so much!

**Reply by broskees** [reply]
@flauc Hey there, trying to implement this - where does this configuration go? Do you only need the JS or do I set up rauthy?

**Comment by ririko5834** [comment]
Hello can you please add private repo support?

**Reply by AnandChowdhary** [reply]
Please see the conversation. You can use @upptime with a private repository.

**Reply by thomasmerz** [reply]
Well, but this only do work with **non-free private repositories**, doesn't it? 🤔

**Reply by AnandChowdhary** [reply]
I don't see why. It should work well with free private repositories as well.

**Reply by thomasmerz** [reply]
Because of this (my repo is private, I'm using a free account on GitHub):

<img width="878" alt="image" src="https://user-images.githubusercontent.com/18568381/106276460-e8007700-6237-11eb-91ae-b68898f1a223.png">

Regarding https://upptime.js.org/docs/get-started/#enable-publishing I can't change "Source" = "None" to gh-pages 🤔

**Reply by AnandChowdhary** [reply]
Aaah, got it! I didn't know that, thanks for pointing it out! 😄

This means that you can use Upptime with free private repos, but without the status page. You still have access to the uptime monitoring and the README.md summary. To enable the status website, you need a Pro plan or a public repo.

**Reply by thomasmerz** [reply]
👍🏻 for clarification! Thanks!
😞 for having no status page with "privacy" (not your "fault")…

**Reply by ririko5834** [reply]
Ok

**Comment by broskees** [comment]
@AnandChowdhary - hoping you have a moment to help me out here.

Ok this is what I did, still not working. I setup a netlify site that basically starts a node server on port 3000... So on deploy it runs `npm install && npm run-script server`

`npm run-script server` just runs `node index.js`.

Then index.js contains:
```js
const express = require('express');
const createProxyMiddleware = require('http-proxy-middleware');

const app = express();

app.use('/uptime/[my_org_name]', createProxyMiddleware({
  target: `https://raw.githubusercontent.com`,
  changeOrigin: true,
  pathRewrite: {[`^/uptime/`]: '/'},
  onProxyReq(proxyReq, req, res) {
    proxyReq.setHeader('Authorization', 'Bearer [my_qhp]');
  }
}));

app.use('/uptime/repos', createProxyMiddleware({
  target: `https://api.github.com`,
  changeOrigin: true,
  pathRewrite: {[`^/uptime/`]: '/'},
  onProxyReq(proxyReq, req, res) {
    proxyReq.setHeader('Authorization', 'Bearer [my_ghp]');
  }
}));

app.listen(3000, () => {
  console.log('Server listening on port ', 3000);
});
```

The reason I placed `uptime` in various places there is because my repo where upptime is ran is called uptime.

In .upptimerc.yml i have:
```yaml
status-website:
  cname: status.mywebsite.com
  publish: true
  apiBaseUrl: my-netlify-app-identifier.netlify.app
```

This isn't doing anything for me, can you tell me what I'm doing wrong?

**Reply by mig5** [reply]
@broskees in case it helps you, here is my working setup.

In all the below, MYORGANIZATION has to match up each time, as does wherever `mydomain.com` occurs. Note also the use of `/status-board` in both the uptime.js as well as the .upptimerc.yml . In my experience, it was the alignment of all these values that make https://uptime.mydomain.com work for me with a private repo.

The uptime.js nodejs app:

```
const express = require('express');
const { createProxyMiddleware } = require('http-proxy-middleware');

const app = express();

app.use('/status-board/MYORGANIZATION', createProxyMiddleware({
  target: 'https://raw.githubusercontent.com',
  changeOrigin: true,
  pathRewrite: {['^/status-board/']: '/'},
  onProxyReq(proxyReq, req, res) {
    proxyReq.setHeader('Authorization', 'Bearer ghp_xxxxxxxxxxxxxxxxxxxxxxxxxxxxx');
  }
}));

app.use('/status-board/repos', createProxyMiddleware({
  target: 'https://api.github.com',
  changeOrigin: true,
  pathRewrite: {['^/status-board/']: '/'},
  onProxyReq(proxyReq, req, res) {
    proxyReq.setHeader('Authorization', 'Bearer ghp_xxxxxxxxxxxxxxxxxxxxxxxxxxxxx');
  }
}));

app.listen(3000, 'localhost', () => {
  console.log('Server listening on port ', 3000);
});
```


The nginx vhost that sits in front:

```
server {
  listen 443;
  server_name uptime.mydomain.com;

  ssl on;
  ssl_certificate           xxxxxxxxxxxxx;
  ssl_certificate_key   xxxxxxxxxxxxx;


  error_log /var/log/nginx/uptime.log info;
  # Session resumption
  ssl_session_cache shared:SSL:10m;
  ssl_session_timeout 10m;

  location / {
        allow 1.2.3.4;
        deny all;
        proxy_pass http://127.0.0.1:3000;
    }
}
```


And my snippet from the .upptimerc.yml:

```
# Change these first
owner: MYORGANIZATION # Your GitHub username
repo: upptime # Your GitHub repository name

# Add your sites here
sites:
[...]

status-website:
  # Uncomment the following line if you don't have a custom domain and add your repo name
  baseUrl: /upptime
  apiBaseUrl: https://uptime.mydomain.com/status-board
  logoUrl: https://raw.githubusercontent.com/upptime/upptime.js.org/master/static/img/icon.svg
  name: MYORGANIZATION Platforms Uptime Monitoring
  introTitle: "**Upptime** is the open-source uptime monitor and status page, powered entirely by GitHub."
  introMessage: Unified real-time uptime reporting, managed via our [GitHub repository](https://github.com/MYORGANIZATION/upptime).
  navbar:
    - title: Status
      href: /
    - title: GitHub
      href: https://github.com/$OWNER/$REPO
  publish: true
  ```
```

**Reply by broskees** [reply]
@mig5 where do you set up the nginx vhost? on the proxy server?

**Reply by mig5** [reply]
@broskees yes - the Nginx server is running on the same machine that runs the nodejs app. (See how it is proxying to 127.0.0.1:3000 which is the nodejs app running on localhost) `proxy_pass http://127.0.0.1:3000;`)

**Reply by broskees** [reply]
@mig5 any idea why I would get a `Cannot Get /` error? Browser console logs indicate CORs issue too.

**Comment by tbringuier** [comment]
Will you add a support without any proxy servers in the future for private repositories ?

**Reply by ScottGuymer** [reply]
I am looking to use this as an internal tool as part of our private github org. It will only show uptime of internal services and only be accessed by people who are a member of the GH org.

I am thinking that the best way to achieve this is the use of a GH app and asking the user to login to the page as a github user using oauth. Their token can then be used to retrieve all of the required info from the API using an authenticated request.

**Reply by ScottGuymer** [reply]
Looks like implicit grant is no longer available so there is a need for a backend to process the token exchange as it requires the client secret..

Looks like that would not be easily possible...

**Comment by sagarkhanal07** [comment]
Asking to confirm if this method is still working on private repo for somebody else?  It is not working for me with the mentioned steps above.

**Reply by IanaRedko** [reply]
does not work for me either

**Comment by PierrickLozach** [comment]
I can't find a page that explains how to configure an API proxy. I see the messages above about using Netlify. Anyone has a tutorial we could user? The documentation only talks about setting apiBaseUrl. Thanks!

**Comment by ronigarcia** [comment]
My status page can't show the graphs in the private repository. All actions were completed successfully, but the status page didn't work.
My upptimerc.yml:
```
status-website:
  cname: upptime-poc.mydomain.com
  publish: true
  apiBaseUrl: ${{ secrets.GH_PAT }}
  baseUrl: /upptime-poc
  logoUrl: https://raw.githubusercontent.com/upptime/upptime.js.org/master/static/img/icon.svg
  name: Upptime
  introTitle: "**Upptime** is the open-source uptime monitor and status page, powered entirely by GitHub."
  introMessage: This is poc status page 
  navbar:
    - title: Status
      href: /
    - title: GitHub
      href: https://github.com/$OWNER/$REPO
 ```

  Could you please help me to find the problem?

**Reply by Rohitpatil-123** [reply]
@ronigarcia Did you manage to figure out the root cause or a workaround for this?

**Comment by ScottGuymer** [comment]
In the end we decided to start using our backstage instance for showing info from our private upptime repo.

We built and released a backstage plugin for upptime here https://github.com/philips-software/philips-backstage-plugins/tree/main/workspaces/upptime/plugins/upptime-backend
